sphinx/source/conf.py

```diff
@@ -201,3 +208,65 @@
 
 # If true, `todo` and `todoList` produce output, else they produce nothing.
 todo_include_todos = True
+
+
+# -- Docstring Improvements --------------------------------------------------
+
+
+def autodoc_process_docstring(
+    app, what: str, name: str, obj, options, lines: List[str]
+) -> None:
+    """
+    Modify docstrings before creating html files.
+
+    Sphinx converts the 'Args:' and 'Returns:' sections of docstrings into
+    reStructuredText (rST) syntax, which can then be found via ':type' & ':rtype'.
+
+    See here for more information:
+    https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html
+    """
+    for i in range(len(lines)):
+        # Skip unless line is an parameter doc or a return doc
+        if not (lines[i].startswith(":type") or lines[i].startswith(":rtype")):
+            continue
+
+        # Change "nn.Module" to "torch.nn.Module" in doc type hints for intersphinx
+        lines[i] = re.sub(r"\bnn\.Module\b", "torch.nn.Module", lines[i])
+        lines[i] = lines[i].replace("torch.torch.", "torch.")
+
+        # Ensure nn.Module and torch.Tensor are hyperlinked
+        lines[i] = re.sub(r"\btorch\.nn\.Module\b", ":obj:`torch.nn.Module`", lines[i])
+        lines[i] = re.sub(r"\btorch\.Tensor\b", ":obj:`torch.Tensor`", lines[i])
+
+        # Handle Any & Callable types
+        lines[i] = re.sub(r"\bAny\b", ":obj:`Any <typing.Any>`", lines[i])
+        lines[i] = re.sub(
+            r"\bCallable\b", ":obj:`Callable <typing.Callable>`", lines[i]
+        )
+
+        # Handle list & tuple types
+        lines[i] = re.sub(r"\blist\b", ":obj:`list`", lines[i])
+        lines[i] = re.sub(r"\btuple\b", ":obj:`tuple`", lines[i])
+
+        # Handle str, bool, & slice types
+        lines[i] = re.sub(r"\bstr\b", ":obj:`str`", lines[i])
+        lines[i] = re.sub(r"\bbool\b", ":obj:`bool`", lines[i])
+        lines[i] = re.sub(r"\bslice\b", ":obj:`slice`", lines[i])
+
+        # Handle int & float types
+        lines[i] = re.sub(r"\bint\b", ":obj:`int`", lines[i])
+        lines[i] = re.sub(r"\bfloat\b", ":obj:`float`", lines[i])
```

Comment:
why do need to manually handle these types which seems to be correctly handled by sphinx

Reply:
@aobo-y Sphinx's automatic type detection cannot handle cases like `list of int`, as it misses the `int` part for example. So we need to explicitly tell Sphinx that these are types that require hyperlinking.

It also fails in other cases as well like this:

![int_float_example](https://user-images.githubusercontent.com/10626398/181404869-269e4c9a-1617-4a3f-9168-3f312fa99b2d.jpg)